plutus-core/testlib/PlutusCore/Generators/PIR/GenerateTypes.hs

```diff
@@ -0,0 +1,501 @@
+                         -- Here we shrink the argument and fixup the function to have the right kind.
+                         [(k, TyApp () (fixKind ctx f (KindArrow () ka' k)) a)
+                         | (ka', a) <- shrinkKindAndType ctx (ka, a)]
+      where ka = unsafeInferKind ctx a
+    -- type lambdas shrink by either shrinking the kind of the argument or shrinking the body
+    TyLam _ x ka b    -> [ (KindArrow () ka' kb, TyLam () x ka' $ substClosedType x (minimalType ka) b)
+                         | ka' <- shrink ka] ++
+                         [ (KindArrow () ka kb', TyLam () x ka b)
+                         | (kb', b) <- shrinkKindAndType (Map.insert x ka ctx) (kb, b)]
+      where KindArrow _ _ kb = k
+    TyForall _ x ka b -> [ (k, b) | not $ Set.member x (ftvTy b) ] ++
+                         -- (above) If the bound variable doesn't matter we get rid of the binding
+                         [ (k, TyForall () x ka' $ substClosedType x (minimalType ka) b)
+                         | ka' <- shrink ka] ++
+                         -- (above) we can always just shrink the bound variable to a smaller kind
+                         -- and ignore it
+                         [ (k, TyForall () x ka b)
+                         | (_, b) <- shrinkKindAndType (Map.insert x ka ctx) (Star, b)]
+                         -- (above) or we shrink the body
+    TyBuiltin{}       -> []
+    TyIFix{}          -> error "shrinkKindAndType: TyIFix"
+
+-- CODE REVIEW: does this exist anywhere?
+inferKind :: Map TyName (Kind ()) -> Type TyName DefaultUni () -> Maybe (Kind ())
+inferKind ctx ty = case ty of
+  TyVar _ x        -> Map.lookup x ctx
+  TyFun _ _ _      -> pure $ Star
+  TyApp _ a _      -> do KindArrow _ _ k <- inferKind ctx a; pure k
+  TyLam _ x k b    -> KindArrow () k <$> inferKind (Map.insert x k ctx) b
+  TyForall _ _ _ _ -> pure $ Star
+  TyBuiltin _ b    -> pure $ builtinKind b
+  TyIFix{}         -> error "inferKind: TyIFix"
+
+-- | Partial unsafeInferKind, useful for context where invariants are set up to guarantee
+-- that types are well-kinded.
+unsafeInferKind :: HasCallStack => Map TyName (Kind ()) -> Type TyName DefaultUni () -> Kind ()
+unsafeInferKind ctx ty =
+  case inferKind ctx ty of
+    Nothing -> error "inferKind"
+    Just k  -> k
+
+-- | Shrink a type in a context assuming that it is of kind *.
+shrinkType :: HasCallStack
+           => Map TyName (Kind ())
+           -> Type TyName DefaultUni ()
+           -> [Type TyName DefaultUni ()]
+shrinkType ctx ty = map snd $ shrinkKindAndType ctx (Star, ty)
+
+-- | Shrink a type of a given kind in a given context in a way that keeps its kind
+shrinkTypeAtKind :: HasCallStack
+                 => Map TyName (Kind ())
+                 -> Kind ()
+                 -> Type TyName DefaultUni ()
+                 -> [Type TyName DefaultUni ()]
+shrinkTypeAtKind ctx k ty = [ ty' | (k', ty') <- shrinkKindAndType ctx (k, ty), k == k' ]
```

Comment:
How likely is this going to give us an empty list?

Reply:
Not entirely unlikely. I can't recall but there used to be a property to measure this. I'll see if I can set up some measurement properties.

Reply:
Why do you ask?